File: python/raspberrypi/DFRobot_RGBLCD1602.py
# -*- coding: utf-8 -*-
''' file DFRobot_RGBLCD1602.cpp
  # DFRobot_RGBLCD1602 class infrastructure, the implementation of basic methods
  @copyright	Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @licence     The MIT License (MIT)
  @maintainer [yangfeng](feng.yang@dfrobot.com)
  @version  V1.0
  @date  2021-09-24
  @url https://github.com/DFRobot/DFRobot_RGBLCD1602
'''
import time
import sys
import wiringpi

LCD_ADDRESS   =  (0x7c>>1)

#color define
WHITE      =     0
RED        =     1
GREEN      =     2
BLUE       =     3
REG_MODE1  =     0x00
REG_MODE2  =     0x01
REG_OUTPUT =     0x08
LCD_CLEARDISPLAY = 0x01
LCD_RETURNHOME = 0x02
LCD_ENTRYMODESET = 0x04
LCD_DISPLAYCONTROL = 0x08
LCD_CURSORSHIFT = 0x10
LCD_FUNCTIONSET = 0x20
LCD_SETCGRAMADDR = 0x40
LCD_SETDDRAMADDR = 0x80

#flags for display entry mode
LCD_ENTRYRIGHT = 0x00
LCD_ENTRYLEFT = 0x02
LCD_ENTRYSHIFTINCREMENT = 0x01
LCD_ENTRYSHIFTDECREMENT = 0x00

#flags for display on/off control
LCD_DISPLAYON = 0x04
LCD_DISPLAYOFF = 0x00
LCD_CURSORON = 0x02
LCD_CURSOROFF = 0x00
LCD_BLINKON = 0x01
LCD_BLINKOFF = 0x00

#flags for display/cursor shift
LCD_DISPLAYMOVE = 0x08
LCD_CURSORMOVE = 0x00
LCD_MOVERIGHT = 0x04
LCD_MOVELEFT = 0x00

#flags for function set
LCD_8BITMODE = 0x10
LCD_4BITMODE = 0x00
LCD_2LINE = 0x08
LCD_1LINE = 0x00
LCD_5x10DOTS = 0x04
LCD_5x8DOTS = 0x00



class DFRobot_RGBLCD1602:
  def __init__(self, rgb_addr,col, row):
    #self.i2c=i2c
    self._row = row
    self._col = col
    self.RGB_ADDRESS = rgb_addr
    print("LCD _row=%d _col=%d"%(self._row,self._col))
    self.LCD = wiringpi.wiringPiI2CSetup(LCD_ADDRESS)
    if self.RGB_ADDRESS == 0x60:
      self.RGB = wiringpi.wiringPiI2CSetup(0x60)
      wiringpi.wiringPiI2CWriteReg8(self.RGB,REG_MODE1, 1)
      self.REG_RED    =     0x04      
      self.REG_GREEN  =     0x03
      self.REG_BLUE   =     0x02
      self.REG_ONLY   =     0x02
    elif self.RGB_ADDRESS == (0x60>>1) :
      self.RGB = wiringpi.wiringPiI2CSetup((0x60>>1))
      self.REG_RED    =     0x06
      self.REG_GREEN  =     0x07
      self.REG_BLUE   =     0x08
      self.REG_ONLY   =     0x08
    elif self.RGB_ADDRESS == (0x6B) :
      self.RGB = wiringpi.wiringPiI2CSetup(0x6B)
      self.REG_RED    =     0x06
      self.REG_GREEN  =     0x05
      self.REG_BLUE   =     0x04
      self.REG_ONLY   =     0x04
    self._show_function = LCD_4BITMODE | LCD_1LINE | LCD_5x8DOTS
    self._begin(self._row,self._col)



  '''
    @brief write character
    @param data the written data
  '''
  def write(self,data):
    b=bytearray(2)
    b[0]=0x40
    b[1]=data
    wiringpi.wiringPiI2CWriteReg8(self.LCD,0x40,data)

  '''
    @brief set RGB
    @param r  red   range(0-255)
    @param g  green range(0-255)
    @param b  blue  range(0-255)
  '''
  def set_RGB(self,r,g,b):
    if self.RGB_ADDRESS == (0x60>>1):
      r=int(r*192/255)
      g=int(g*192/255)
      b=int(b*192/255)
    self._set_reg(self.REG_RED,r)
    self._set_reg(self.REG_GREEN,g)
    self._set_reg(self.REG_BLUE,b)
    if self.RGB_ADDRESS == 0x6b:
      self._set_reg(0x07,0xFF)
  '''
    @brief set cursor position
    @param col columns optional range 0-15
    @param row rows optional range 0-1，0 is the first row, 1 is the second row
  '''
  def set_cursor(self,col,row):
    if(row == 0):
      col|=0x80
    else:
      col|=0xc0
    self._command(col)

  '''
    @brief clear the display and return the cursor to the initial position (position 0)
  '''
  def clear(self):
    self._command(LCD_CLEARDISPLAY)
    time.sleep(0.002)

  '''
    @brief scroll left to display
  '''
  def scroll_display_left(self):
    self._command(LCD_CURSORSHIFT | LCD_DISPLAYMOVE | LCD_MOVELEFT)
  '''
    @brief scroll right to display
  '''
  def scroll_display_right(self):
    self._command(LCD_CURSORSHIFT | LCD_DISPLAYMOVE | LCD_MOVERIGHT)

  '''
    @brief output data to LCD to display
    @param arg output data
  '''
  def print_out(self,arg):
    if(isinstance(arg,int)):
      arg=str(arg)

    for x in bytearray(arg,'utf-8'):
      self.write(x)

  '''
    @brief return the cursor to the initial position（0,0）
  '''
  def home(self):
    self._command(LCD_RETURNHOME)        # set cursor position to zero
    time.sleep(1)        # this _command takes a long time!

    '''
      @brief Turn off the display
    '''
  def no_display(self):
    self._show_control &= ~LCD_DISPLAYON 
    self._command(LCD_DISPLAYCONTROL | self._show_control)

  '''
    @brief Turn on the display
  '''
  def display(self):
    self._show_control |= LCD_DISPLAYON 
    self._command(LCD_DISPLAYCONTROL | self._show_control)

  '''
    @brief Turn  off the blinking showCursor
  '''
  def stop_blink(self):
    self._show_control &= ~LCD_BLINKON 
    self._command(LCD_DISPLAYCONTROL | self._show_control)

  '''
    @brief Turn on  the blinking showCursor
  '''
  def blink(self):
    self._show_control |= LCD_BLINKON 
    self._command(LCD_DISPLAYCONTROL | self._show_control)

  '''
    @brief Turn off the underline showCursor 
  '''
  def no_cursor(self):
    self._show_control &= ~LCD_CURSORON 
    self._command(LCD_DISPLAYCONTROL | self._show_control)

  '''
    @brief Turn on the underline showCursor 
  '''
  def cursor(self):
    self._show_control |= LCD_CURSORON 
    self._command(LCD_DISPLAYCONTROL | self._show_control)

  '''
    @brief This is for text that flows Left to Right
  '''
  def left_to_right(self):
    self._show_mode |= LCD_ENTRYLEFT 
    self._command(LCD_ENTRYMODESET | self._show_mode)

  '''
    @brief This is for text that flows Right to Left
  '''
  def right_to_left(self):
    self._show_mode &= ~LCD_ENTRYLEFT 
    self._command(LCD_ENTRYMODESET | self._show_mode)

  '''
    @brief This will 'left justify' text from the showCursor
  '''
  def no_autoscroll(self):
    self._show_mode &= ~LCD_ENTRYSHIFTINCREMENT 
    self._command(LCD_ENTRYMODESET | self._show_mode)

  '''
    @brief This will 'right justify' text from the showCursor
  '''
  def autoscroll(self):
    self._show_mode |= LCD_ENTRYSHIFTINCREMENT 
    self._command(LCD_ENTRYMODESET | self._show_mode)

  '''
    @brief Allows us to fill the first 8 C
    @param location substitute character range（0-7）
    @param charmap  character listing the size is 8 bytes
  '''
  def customSymbol(self,location, charmap):
    location &= 0x7  # we only have 8 locations 0-7
    self._command(LCD_SETCGRAMADDR | (location << 3))
    for i in range(0,8):
      wiringpi.wiringPiI2CWriteReg8(self.LCD,0x40,charmap[i])

  '''
    @brief set the backlight
    @param mode  true indicates the backlight is turned on and set to white, false indicates the backlight is turned off
  '''
  def setBacklight(self,mode):
    if(mode):
      self.set_color_white()      # turn backlight on
    else:
      self.close_backlight()        # turn backlight off

  '''
    @brief output string
  '''
  def printstr(self,c):
    #/< This function is not identical to the function used for "real" I2C displays
    #/< it's here so the user sketch doesn't have to be changed 
    self.print_out(c)
 
  '''
     @brief set backlight PWM output
     @param color  backlight color  Preferences：REG_RED\REG_GREEN\REG_BLUE
     @param pwm  color intensity   range(0-255)
  '''
  def set_pwm(self,color,pwm):
    self._set_reg(color, pwm)
    if self.RGB_ADDRESS == 0x6b:
      self._set_reg(0x07,0xFF)

  '''
     @brief set the backlight to white
  '''
  def set_color_white(self):
    self.set_RGB(255, 255, 255)

  '''
    @brief close the backlight
  '''
  def close_backlight(self):
    self.set_RGB(0, 0, 0)

  '''
    @brief the initialization function
    @param cols columns optional range 0-15
    @param lines rows optional range 0-1，0 is the first row, 1 is the second row
    @param dotsize  character size LCD_5x8DOTS\LCD_5x10DOTS
  '''
  def _begin(self,cols,lines,dotsize=LCD_5x8DOTS):
    if (lines > 1):
        self._show_function |= LCD_2LINE 
     
    self._num_lines = lines 
    self._curr_line = 0 

    # for some 1 line displays you can select a 10 pixel high font
    if ((dotsize != 0) and (lines == 1)) :
        self._show_function |= LCD_5x10DOTS 
     
    # SEE PAGE 45/46 FOR INITIALIZATION SPECIFICATION!
    # according to datasheet, we need at least 40ms after power rises above 2.7V
    # before sending _commands. Arduino can turn on way befer 4.5V so we'll wait 50
    #delayMicroseconds(50000);
    time.sleep(0.05)

    # this is according to the hitachi HD44780 datasheet
    # page 45 figure 23

    # Send function set _command sequence
    self._command(LCD_FUNCTIONSET | self._show_function)
    #delayMicroseconds(4500);  # wait more than 4.1ms
    time.sleep(0.005)
    # second try
    self._command(LCD_FUNCTIONSET | self._show_function)
    #delayMicroseconds(150);
    time.sleep(0.005)
    # third go
    self._command(LCD_FUNCTIONSET | self._show_function)
    # finally, set # lines, font size, etc.
    self._command(LCD_FUNCTIONSET | self._show_function)
    # turn the display on with no cursor or blinking default
    self._show_control = LCD_DISPLAYON | LCD_CURSOROFF | LCD_BLINKOFF 
    self.display()
    # clear it off
    self.clear()
    # Initialize to default text direction (for romance languages)
    self._show_mode = LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT 
    # set the entry mode
    self._command(LCD_ENTRYMODESET | self._show_mode)
    if self.RGB_ADDRESS == 0x60:
      # backlight init
      self._set_reg(REG_MODE1, 0)
      # set LEDs controllable by both PWM and GRPPWM registers
      self._set_reg(REG_OUTPUT, 0xFF)
      # set MODE2 values
      # 0010 0000 -> 0x20  (DMBLNK to 1, ie blinky mode)
      self._set_reg(REG_MODE2, 0x20)
    elif self.RGB_ADDRESS == (0x60>>1) :
      self._set_reg(0x04, 0x15)
    elif self.RGB_ADDRESS == (0x6B) :
      self._set_reg(0x2F, 0x00)
      self._set_reg(0x00, 0x20)
      self._set_reg(0x01, 0x00)
      self._set_reg(0x02, 0x01)
      self._set_reg(0x03, 0x04)
    self.set_color_white()

  '''
    @brief send command
    @param data the sent data
  '''
  def _command(self,cmd):
    b=bytearray(2)
    b[0]=0x80
    b[1]=cmd
    wiringpi.wiringPiI2CWriteReg8(self.LCD,0x80,cmd)

  '''
    @brief set the register
    @param addr register address
    @param data data
  '''
  def _set_reg(self,reg,data):
    b=bytearray(1)
    b[0]=data
    wiringpi.wiringPiI2CWriteReg8(self.RGB,reg,data)

File: DFRobot_RGBLCD1602.cpp
/*!
 * @file DFRobot_RGBLCD1602.cpp
 * @brief DFRobot_RGBLCD1602 class infrastructure, the implementation of basic methods
 * @copyright	Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @licence     The MIT License (MIT)
 * @maintainer [yangfeng](feng.yang@dfrobot.com)
 * @version  V1.0
 * @date  2021-09-24
 * @url https://github.com/DFRobot/DFRobot_RGBLCD1602
 */

#include <stdio.h>
#include <string.h>


#include "DFRobot_RGBLCD1602.h"

const uint8_t color_define[4][3] = 
{
    {255, 255, 255},            // white
    {255, 0, 0},                // red
    {0, 255, 0},                // green
    {0, 0, 255},                // blue
};

/*******************************public*******************************/
DFRobot_RGBLCD1602::DFRobot_RGBLCD1602(uint8_t RGBAddr,uint8_t lcdCols,uint8_t lcdRows,TwoWire *pWire,uint8_t lcdAddr)
{
  _lcdAddr = lcdAddr;
  _RGBAddr = RGBAddr;
  _cols = lcdCols;
  _rows = lcdRows;
  _pWire = pWire;
}

void DFRobot_RGBLCD1602::init()
{
  _pWire->begin();
  if(_RGBAddr == (0x60)){
    REG_RED   =      0x04;
    REG_GREEN =      0x03;
    REG_BLUE  =      0x02;
    REG_ONLY  =      0x02 ;
  } else if(_RGBAddr == (0x60>>1)){
    REG_RED      =   0x06 ;       // pwm2
    REG_GREEN    =   0x07 ;       // pwm1
    REG_BLUE     =   0x08 ;       // pwm0
    REG_ONLY     =   0x08 ;
  } else if(_RGBAddr == (0x6B)){
    REG_RED      =   0x06 ;       // pwm2
    REG_GREEN    =   0x05 ;       // pwm1
    REG_BLUE     =   0x04 ;       // pwm0
    REG_ONLY     =   0x04 ; 
  } else if(_RGBAddr == (0x2D)){
    REG_RED      =   0x01 ;       // pwm2
    REG_GREEN    =   0x02 ;       // pwm1
    REG_BLUE     =   0x03 ;       // pwm0
    REG_ONLY     =   0x01 ; 
  }
  _showFunction = LCD_4BITMODE | LCD_1LINE | LCD_5x8DOTS;
  begin(_rows);
}

void DFRobot_RGBLCD1602::clear()
{
    command(LCD_CLEARDISPLAY);        // clear display, set cursor position to zero
    delayMicroseconds(2000);          // this command takes a long time!
}

void DFRobot_RGBLCD1602::home()
{
    command(LCD_RETURNHOME);        // set cursor position to zero
    delayMicroseconds(2000);        // this command takes a long time!
}

void DFRobot_RGBLCD1602::noDisplay()
{
    _showControl &= ~LCD_DISPLAYON;
    command(LCD_DISPLAYCONTROL | _showControl);
}

void DFRobot_RGBLCD1602::display() {
    _showControl |= LCD_DISPLAYON;
    command(LCD_DISPLAYCONTROL | _showControl);
}

void DFRobot_RGBLCD1602::stopBlink()
{
    _showControl &= ~LCD_BLINKON;
    command(LCD_DISPLAYCONTROL | _showControl);
}
void DFRobot_RGBLCD1602::blink()
{
    _showControl |= LCD_BLINKON;
    command(LCD_DISPLAYCONTROL | _showControl);
}

void DFRobot_RGBLCD1602::noCursor()
{
    _showControl &= ~LCD_CURSORON;
    command(LCD_DISPLAYCONTROL | _showControl);
}

void DFRobot_RGBLCD1602::cursor() {
    _showControl |= LCD_CURSORON;
    command(LCD_DISPLAYCONTROL | _showControl);
}

void DFRobot_RGBLCD1602::scrollDisplayLeft(void)
{
    command(LCD_CURSORSHIFT | LCD_DISPLAYMOVE | LCD_MOVELEFT);
}

void DFRobot_RGBLCD1602::scrollDisplayRight(void)
{
    command(LCD_CURSORSHIFT | LCD_DISPLAYMOVE | LCD_MOVERIGHT);
}

void DFRobot_RGBLCD1602::leftToRight(void)
{
    _showMode |= LCD_ENTRYLEFT;
    command(LCD_ENTRYMODESET | _showMode);
}

void DFRobot_RGBLCD1602::rightToLeft(void)
{
    _showMode &= ~LCD_ENTRYLEFT;
    command(LCD_ENTRYMODESET | _showMode);
}

void DFRobot_RGBLCD1602::noAutoscroll(void)
{
    _showMode &= ~LCD_ENTRYSHIFTINCREMENT;
    command(LCD_ENTRYMODESET | _showMode);
}

void DFRobot_RGBLCD1602::autoscroll(void)
{
    _showMode |= LCD_ENTRYSHIFTINCREMENT;
    command(LCD_ENTRYMODESET | _showMode);
}

void DFRobot_RGBLCD1602::customSymbol(uint8_t location, uint8_t charmap[])
{

    location &= 0x7; // we only have 8 locations 0-7
    command(LCD_SETCGRAMADDR | (location << 3));
    
    
    uint8_t data[9];
    data[0] = 0x40;
    for(int i=0; i<8; i++)
    {
        data[i+1] = charmap[i];
    }
    send(data, 9);
}

void DFRobot_RGBLCD1602::setCursor(uint8_t col, uint8_t row)
{

    col = (row == 0 ? col|0x80 : col|0xc0);
    uint8_t data[3] = {0x80, col};

    send(data, 2);

}

void DFRobot_RGBLCD1602::setRGB(uint8_t r, uint8_t g, uint8_t b)
{
  uint16_t temp_r,temp_g,temp_b;
  if(_RGBAddr == 0x60>>1){
    temp_r = (uint16_t)r*192/255;
    temp_g = (uint16_t)g*192/255;
    temp_b = (uint16_t)b*192/255;
    setReg(REG_RED, temp_r);
    setReg(REG_GREEN, temp_g);
    setReg(REG_BLUE, temp_b);
  } else{
    setReg(REG_RED, r);
    setReg(REG_GREEN, g);
    setReg(REG_BLUE, b);
    if(_RGBAddr == 0x6B){
      setReg(0x07, 0xFF);
    }
  }

}

void DFRobot_RGBLCD1602::setColor(uint8_t color)
{
    if(color > 3)return ;
    setRGB(color_define[color][0], color_define[color][1], color_define[color][2]);
}


inline size_t DFRobot_RGBLCD1602::write(uint8_t value)
{

    uint8_t data[3] = {0x40, value};
    send(data, 2);
    return 1; // assume sucess
}

inline void DFRobot_RGBLCD1602::command(uint8_t value)
{
    uint8_t data[3] = {0x80, value};
    send(data, 2);
}



void DFRobot_RGBLCD1602::setBacklight(bool mode){
	if(mode){
		setColorWhite();		// turn backlight on
	}else{
		closeBacklight();		// turn backlight off
	}
}

/*******************************private*******************************/
void DFRobot_RGBLCD1602::begin( uint8_t rows, uint8_t charSize) 
{
    if (rows > 1) {
        _showFunction |= LCD_2LINE;
    }
    _numLines = rows;
    _currLine = 0;
    ///< for some 1 line displays you can select a 10 pixel high font
    if ((charSize != 0) && (rows == 1)) {
        _showFunction |= LCD_5x10DOTS;
    }

    ///< SEE PAGE 45/46 FOR INITIALIZATION SPECIFICATION!
    ///< according to datasheet, we need at least 40ms after power rises above 2.7V
    ///< before sending commands. Arduino can turn on way befer 4.5V so we'll wait 50
    delay(50);

    ///< this is according to the hitachi HD44780 datasheet
    ///< page 45 figure 23

    ///< Send function set command sequence
    command(LCD_FUNCTIONSET | _showFunction);
    delay(5);  // wait more than 4.1ms
	
	///< second try
    command(LCD_FUNCTIONSET | _showFunction);
    delay(5);

    ///< third go
    command(LCD_FUNCTIONSET | _showFunction);

    ///< turn the display on with no cursor or blinking default
    _showControl = LCD_DISPLAYON | LCD_CURSOROFF | LCD_BLINKOFF;
    display();

    ///< clear it off
    clear();

    ///< Initialize to default text direction (for romance languages)
    _showMode = LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT;
    ///< set the entry mode
    command(LCD_ENTRYMODESET | _showMode);
    
    if(_RGBAddr == (0xc0>>1)){
      ///< backlight init
      setReg(REG_MODE1, 0);
      ///< set LEDs controllable by both PWM and GRPPWM registers
      setReg(REG_OUTPUT, 0xFF);
      ///< set MODE2 values
      ///< 0010 0000 -> 0x20  (DMBLNK to 1, ie blinky mode)
      setReg(REG_MODE2, 0x20);
    }else if(_RGBAddr == (0x60>>1)){
       setReg(0x01, 0x00);
       setReg(0x02, 0xfF);
       setReg(0x04, 0x15);
    }else if(_RGBAddr==0x6B){
        setReg(0x2F, 0x00);
        setReg(0x00, 0x20);
        setReg(0x01, 0x00);
        setReg(0x02, 0x01);
        setReg(0x03, 4);
    }
    setColorWhite();
}

void DFRobot_RGBLCD1602::send(uint8_t *data, uint8_t len)
{
    _pWire->beginTransmission(_lcdAddr);        // transmit to device #4
    for(int i=0; i<len; i++) {
        _pWire->write(data[i]);
		delay(5);
    }
    _pWire->endTransmission();                     // stop transmitting
}

void DFRobot_RGBLCD1602::setReg(uint8_t addr, uint8_t data)
{
    _pWire->beginTransmission(_RGBAddr); // transmit to device #4
    _pWire->write(addr);
    _pWire->write(data);
    _pWire->endTransmission();    // stop transmitting
}



